steward_proto/proto/aave_v2_stablecoin.proto

```diff
@@ -1,68 +1,118 @@
+        Sweep sweep = 6;
+        // Represents function `transferFees()`: [link](https://github.com/PeggyJV/cellar-contracts/blob/b93b4393299aee35421f83ef4ca2689256e5b354/contracts/AaveV2StablecoinCellar.sol#L754)
+        TransferFees transfer_fees = 7;
     }
-}
 
-// function enterStrategy() external;
-message EnterStrategy {}
+    /*
+     * Take platform fees and performance fees off of cellar's active assets.
+     *
+     * Represents function `accruePlatformFees()`
+     */
+    message AccrueFees {}
 
-// function reinvest(uint256 amount, uint256 minAssetsOut) external;
-message ReinvestWithAmount {
-    uint64 amount = 1;
-    uint64 min_assets_out = 2;
-}
+    /*
+     * Claim rewards from Aave and begin cooldown period to unstake them.
+     *
+     * Represents function `claimAndUnstake()`
+     */
+    message ClaimAndUnstake {}
 
-// function reinvest(uint256 minAssetsOut) external;
-message Reinvest {
-    uint64 min_assets_out = 1;
-}
+    /*
+     * Enters inactive assets into the current Aave stablecoin position.
+     *
+     * Represents function `enterPosition()`
+     */
+    message EnterPosition {}
 
-// function claimAndUnstake(uint256 amount) external returns (uint256 claimed);
-message ClaimAndUnstakeWithAmount {
-    uint64 amount = 1;
-}
-
-// function claimAndUnstake() external returns (uint256);
-message ClaimAndUnstake {}
+    /*
+     * Rebalances current assets into a new asset position.
+     *
+     * Represents function `rebalance(address newLendingToken, uint256 minNewLendingTokenAmount)`
+     *
+     * This function is based on the Curve Pool Registry exchange_multiple() function:
+     * https://github.com/curvefi/curve-pool-registry/blob/16a8664952cf61d7fed06acca79ad5ac696f4b20/contracts/Swaps.vy#L461-L489
+     */
+    message Rebalance {
+        // An array of up to 9 addresses (4 swaps) representing a token swap route, where each triplet of addresses is a single swap route (ex. in token address, pool address, out token address)
+        repeated string route = 1;
+        // An array of up to 4 swap params. Attempting more than four swaps will fail.
+        repeated SwapParams swap_params = 2;
+        // Minimum acceptable assets to be received from the swap (slippage parameter)
+        uint64 min_assets_out = 3;
 
-// function rebalance(address newLendingToken, uint256 minNewLendingTokenAmount) external;
-message Rebalance {
-    string address = 1;
-    uint64 min_new_lending_token_amount = 2;
-}
-
-// function accruePlatformFees() external;
-message AccruePlatformFees {}
-
-// function transferFees() external;
-message TransferFees {}
+        /*
+         * Represents parameters for a single swap. Each swap needs the indeces in Rebalance.route of the in/out token addresses and the swap type. See the Curve contract linked above for more detail.
+         */
+        message SwapParams {
```

Comment:
It kind of feels ugly our code has to understand the semantics of the particular arguments in order to be able to to parse it and convert it into the expected format by the contract, but there isn't much alternative at the moment absent requiring the SP to format the encoded call and doing inspection of that call rather than creating it. Something to think about.

Reply:
We could expect SPs to do input validation themselves, it kind of makes sense in the case where they are the ones who created the cellar. I wanted to make the UX and validation simpler to avoid wasted gas on mistakes but in reality we probably don't want to be in the business of validating all args... If we put that responsibility on the SP then it would reduce the amount of code we need per cellar as you mentioned elsewhere.